plugins/gossip.js

```diff
@@ -0,0 +1,281 @@
+      peer.disconnect = function (err, cb) {
+        if (isFunction(err)) {
+          cb = err
+          err = null
+        }
+        rpc.close(err, cb)
+      }
+
+      if (isClient) {
+        // default ping is 5 minutes...
+        var pp = ping({serve: true, timeout: timerPing}, function (_) {})
+        peer.ping = {rtt: pp.rtt, skew: pp.skew}
+        pull(
+          pp,
+          rpc.gossip.ping({timeout: timerPing}, function (err) {
+            if (err.name === 'TypeError') peer.ping.fail = true
+          }),
+          pp
+        )
+      }
+
+      rpc.on('closed', function () {
+        server.emit('log:info', ['SBOT', stringify(peer),
+          ['DISCONNECTED. state was', peer.state, 'for',
+            (new Date() - peer.stateChange) / 1000, 'seconds'].join(' ')])
+        // track whether we have successfully connected.
+        // or how many failures there have been.
+        var since = peer.stateChange
+        peer.stateChange = Date.now()
+        //        if(peer.state === 'connected') //may be "disconnecting"
+        peer.duration = stats(peer.duration, peer.stateChange - since)
+        peer.state = undefined
+        notify({ type: 'disconnect', peer: peer })
+      })
+
+      notify({ type: 'connect', peer: peer })
+    })
+
+    var last
+    stateFile.get(function (err, ary) {
+      last = ary || []
+      if (Array.isArray(ary)) {
+        ary.forEach(function (v) {
+          delete v.state
+          // don't add local peers (wait to rediscover)
+          if (v.source !== 'local') {
+            gossip.add(v, 'stored')
+          }
+        })
+      }
+    })
+
+    var int = setInterval(function () {
+      var copy = JSON.parse(JSON.stringify(gossip.peers()))
+      copy.filter(function (e) {
+        return e.source !== 'local'
+      }).forEach(function (e) {
+        delete e.state
+      })
+      if (deepEqual(copy, last)) return
+      last = copy
+      // TODO: what to do with this?
+      // stateFile.set(copy, function (err) {
+      //  if (err) console.log(err)
+      // })
+    }, 10 * 1000)
+
+    if (int.unref) int.unref()
+
+    return gossip
+  }
+
+}
+
+function isFunction (f) {
+  return typeof f === 'function'
+}
+
+function stringify (peer) {
+  return [peer.host, peer.port, peer.key].join(':')
+}
+
+function isObject (o) {
+  return o && typeof o === 'object'
+}
+
+function toBase64 (s) {
+  if (isString(s)) return s
+  else s.toString('base64') // assume a buffer
+}
+
+function isString (s) {
+  return typeof s === 'string'
+}
+
+var feedIdRegex = new RegExp(ref.feedIdRegex)
+
+function toMultiserverAddress (address) {
+  if (isObject(address)) {
+    if (ref.isFeed(address.key)) {
+      address.key = address.key.match(feedIdRegex)[1]
+    }
+    var protocol = 'net'
+    if (address.host.endsWith('.onion')) { protocol = 'onion' }
+    return [protocol, address.host, address.port].join(':') + '~' + ['shs', toBase64(address.key)].join(':')
```

Comment:
This piece of code (I think it's old from the previous code) assumes the existence of port and host in the address object. We should eventually not assume that from every multiserver address.